docs/rfc/0003_risingwave_ng.md

```diff
@@ -0,0 +1,208 @@
+
+# **Explanation**
+
+The RFC mainly adds a new CRD and rewrites the RisingWave CRD. It introduces the `RisingWavePodTemplate` for declaring pod templates, just like the template in Deployment and StatefulSet.
+
+## RisingWavePodTemplate
+
+```yaml
+apiVersion: risingwave.singularity-data.com/v1alpha1
+kind: RisingWavePodTemplate
+metadata:
+  name: template-a
+spec:
+  metadata: # Pod annotations and labels
+  spec:     # Pod Spec, only the first container's spec would be overridden, and the other containers' spec will be copied
+```
+
+`RisingWavePodTemplate` is immutable so can not be changed after it's created. When creating a RisingWave that refers to the template, the Webhook should validate if there's such an object and reject if not. And also, the webhook should add finalizers to the object so that it would not be deleted accidentally and remove the finalizers when RisingWave objects are going to be deleted.
+
+## RisingWave
+
+Here's an example [risingwave_ng.yaml](./examples/risingwave_ng.yaml). It mainly includes these parts:
+
++ `global` for simplified specs, such as image, image pull policy, pod template, resources, node selector and replicas
++ `components` for advanced specs, one can
+  + Declare complex groups of pods of components, specifying different templates, images, resources and other things related
+  + Specifying the listen ports of components
+  + The configs in one `group` should be merged into the globals, e.g.,
+    + maps are merged and overridden if there're new values (resources)
+    + lists are merged
+    + values are overridden
+  + The groups here are extended groups to the global ones, i.e. if there're 2 compactors in global and a group of 2 compactors in components, there will be 4 in total
++ `configuration` for using pre-defined configs
++ `storages` for specifying the storage types and credentials, as well as the PVC template for compute nodes
++ `security` for specifying the TLS configs
+
+Also, the `podTemplate` will be read and applied onto the Pods of RisingWave if it is set.
+
+Here're examples to demonstrate the new CRD:
+
++ 1/1/1/1, 1C1G
+
+```yaml
+apiVersion: risingwave.singularity-data.com/v1alpha1
+kind: RisingWave
+metadata:
+  name: test-risingwave
+spec:
+  global:
+    image: ghcr.io/singularity-data/risingwave:latest
+    replicas:
+      meta: 1
+      frontend: 1
+      compute: 1
+      compactor: 1
+    resources:
+      limits:
+        cpu: 1
+        memory: 1Gi
+  storages:
+    meta:
+      memory: true
+    object:
+      memory: true
+```
+
++ 1/1/1/1, 2 spot compactors on a node group labeled with "spot", 1C1G
+
+```yaml
+apiVersion: risingwave.singularity-data.com/v1alpha1
+kind: RisingWave
+metadata:
+  name: test-risingwave
+spec:
+  # ...
+  components:
+    compactor:
+      groups:
+      - name: spot
+        replicas: 2
+        nodeSelector:
+          node-group: spot
```

Comment:
should it be nodeGroup (cammelCase seems like the convention)?
Should we use affinity instead of nodeSelector?
`affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
            - matchExpressions:
              - key: nodeGroup
                operator: In
                values:
                - spot`

Reply:
> should it be nodeGroup (cammelCase seems like the convention)? Should we use affinity instead of nodeSelector? `affinity: nodeAffinity: requiredDuringSchedulingIgnoredDuringExecution: nodeSelectorTerms: - matchExpressions: - key: nodeGroup operator: In values: - spot`

No, this is not affinity (a structure) but simply node groups (a map of labels). It's less expressive than the `affinity`, but it's enough for node scheduling. IMO The `affinity` things (like scatter/gather among pods) should be considered by the operator rather than by the user. Thus I removed that field from the CRD.

Reply:
By convention, we use dash-linked strings for labels and annotations.